sklearn/metrics/tests/test_pairwise_distances_reduction.py

```diff
@@ -96,24 +148,498 @@ def assert_radius_neighborhood_results_equality(ref_dist, dist, ref_indices, ind
+
+        for j in range(m):
+            rounded_dist = np.round(ref_dist_row[j], decimals=decimals)
+            ref_mapping[rounded_dist].add(ref_indices_row[j])
+            mapping[rounded_dist].add(indices_row[j])
+
+        # Asserting equality of groups (sets) for each distance
+        for j in ref_mapping.keys():
+            assert ref_mapping[j] == mapping[j]
+
+
 ASSERT_RESULT = {
-    PairwiseDistancesArgKmin: assert_argkmin_results_equality,
-    PairwiseDistancesRadiusNeighborhood: assert_radius_neighborhood_results_equality,
+    # In the case of 64bit, we test for exact equality.
+    (PairwiseDistancesArgKmin, np.float64): assert_argkmin_results_equality,
+    (
+        PairwiseDistancesRadiusNeighborhood,
+        np.float64,
+    ): assert_radius_neighborhood_results_equality,
+    # In the case of 32bit, indices can be permuted due to small difference
+    # in the computations of their associated distances, hence we test equality of
+    # results up to valid permutations.
+    (PairwiseDistancesArgKmin, np.float32): assert_argkmin_results_quasi_equality,
+    (
+        PairwiseDistancesRadiusNeighborhood,
+        np.float32,
+    ): assert_radius_neighborhood_results_quasi_equality,
 }
 
 
+def test_assert_argkmin_results_quasi_equality():
+
+    rtol = 1e-7
+    atol = 1e-7
+    decimals = 6
+
+    ref_dist = np.array(
+        [
+            [1.2, 2.5, 6.1 - atol, 6.1, 6.1 + atol],
+            [1.0 - atol, 1.0 - atol, 1, 1.0 + atol, 1.0 + atol],
+        ]
+    )
+    ref_indices = np.array(
+        [
+            [1, 2, 3, 4, 5],
+            [6, 7, 8, 9, 10],
+        ]
+    )
+
+    assert_argkmin_results_quasi_equality(
+        ref_dist, ref_dist, ref_indices, ref_indices, rtol, decimals
+    )
+
+    # Apply valid permutation on indices
+    assert_argkmin_results_quasi_equality(
+        ref_dist=np.array(
+            [
+                [1.2, 2.5, 6.1 - atol, 6.1, 6.1 + atol],
+            ]
+        ),
+        dist=np.array(
+            [
+                [1.2, 2.5, 6.1 - atol, 6.1, 6.1 + atol],
+            ]
+        ),
+        ref_indices=np.array(
+            [
+                [1, 2, 3, 4, 5],
+            ]
+        ),
+        indices=np.array(
+            [
+                [1, 2, 4, 5, 3],
+            ]
+        ),
+        rtol=rtol,
+        decimals=decimals,
+    )
+    assert_argkmin_results_quasi_equality(
+        ref_dist=np.array(
+            [
+                [1.0 - atol, 1.0 - atol, 1, 1.0 + atol, 1.0 + atol],
+            ]
+        ),
+        dist=np.array(
+            [
+                [1.0 - atol, 1.0 - atol, 1, 1.0 + atol, 1.0 + atol],
+            ]
+        ),
+        ref_indices=np.array(
+            [
+                [6, 7, 8, 9, 10],
+            ]
+        ),
+        indices=np.array(
+            [
+                [6, 9, 7, 8, 10],
+            ]
+        ),
+        rtol=rtol,
+        decimals=decimals,
+    )
+
+    # Apply invalid permutation on indices
```

Comment:
```suggestion
    # Apply invalid permutation on indices: permuting the ranks
    # of the 2 nearest neighbors is invalid because the distance
    # values are too different.
```